File: backends/php-laravel/app/Support/ApiProblem.php
<?php

namespace App\Support;

use RuntimeException;

class ApiProblem extends RuntimeException
{
    public function __construct(
        public readonly int $status,
        public readonly string $title,
        public readonly ?string $detail = null,
        public readonly ?string $detailKey = null,
    ) {
        parent::__construct($detail ?? $title);
    }
}

File: backends/spring-java/src/main/java/dev/stackverse/backend/common/ApiProblem.java
package dev.stackverse.backend.common;

import org.springframework.http.HttpStatus;

/**
 * Application exceptions that map 1:1 onto RFC 9457 problem documents.
 */
public class ApiProblem extends RuntimeException {
    private final HttpStatus status;
    private final String title;
    private final String detailKey;
    private final String detail;

    public ApiProblem(HttpStatus status, String title, String detailKey, String detail) {
        super(detail == null ? title : detail);
        this.status = status;
        this.title = title;
        this.detailKey = detailKey;
        this.detail = detail;
    }

    public HttpStatus getStatus() {
        return status;
    }

    public String getTitle() {
        return title;
    }

    public String getDetailKey() {
        return detailKey;
    }

    public String getDetail() {
        return detail;
    }
}
